Implement specialized pool support in the scheduler

mihails-strasuns-sociomantic opened the pull request:
Shows planned design and passes most simple test. PR right now only exists as form of preview and code coverage report abuse.

TODO:

- [x] More comprehensive tests
- [x] Review / fix documentation
- [ ] Test using one of existing applications extensively using tasks

mihails-strasuns-sociomantic committed:
Move most logic from `runTask` to `entryPoint`

This is an inherent part of running a task, not a matter for the context
it is being run from. Moving it from scheduler into task entry point
method will also improve encapsulation of private task fields.

mihails-strasuns-sociomantic committed:
Make 'termination_hooks` private again

Now that code using it is moved into Task module, it can be switched
from "documented private" to actual private.

mihails-strasuns-sociomantic committed:
Split away task queue subsystem of scheduler

Refactoring aimed to reduce complexity of `Scheduler` module by
splitting self-contained independent utilities.

mihails-strasuns-sociomantic committed:
Implement FiberPool derivative able to run tasks

Much more simple alternative to `FiberPoolWithQueue` as it always runs
specified task immediately, growing pool as needed indefinitely.

mihails-strasuns-sociomantic committed:
Implement specialized pool support in the scheduler

gavin-norman-sociomantic commented on src/ocean/task/internal/FiberPoolEager.d:
```diff
@@ -0,0 +1,82 @@
+
+    public void run ( Task task )
+    {
+        auto fiber = this.get();
```
So this is an unlimited pool of fibers, right?

mihails-strasuns-sociomantic replied:
Yes. My line of though went like that - we are only expected to use those pools for real-time kind of handling where waiting in queue is not appropriate. Like HTTP connection handler. If we allow the limit, what would be a reasonable fallback for case when it is reached? Close the connection, bring down the app? Nothing really made sense. It feels like having unbound limit and just monitoring graphs is the only practical approach.

gavin-norman-sociomantic replied:
I see. This works fine for normal usage where the number of connections (say) gradually grows over time. We also need to consider erroneous or malicious behaviour, though, where the number of connections could grow extremely quickly. OOM is not an acceptable solution ;)

gavin-norman-sociomantic replied:
As far as I remember from our discussion on this before, the conclusion was that this is fine/irrelevant, as the select listener already implements a limit on the number of handled connections. (And rejects any incoming connections beyond the limit.)

mihails-strasuns-sociomantic replied:
Correct. Consensus was to think about implementing a limit once we have real app use case that may require it.

gavin-norman-sociomantic commented on src/ocean/task/internal/SpecializedPools.d:
```diff
@@ -0,0 +1,124 @@
+
+    ***************************************************************************/
+
+    bool run ( Task task )
```
Do you anticipate use cases where the caller isn't sure whether the task is handled by this class or not? I would have expected it to throw if the task isn't of a supported type.

mihails-strasuns-sociomantic replied:
Right now it is used to simplify code in scheduler. It does something like `if (!custom_pools.run(task)) generic_pools.run(task);`.

gavin-norman-sociomantic replied:
Fair enough.

gavin-norman-sociomantic reviewed (commented):
Interesting that all specialized task pools are unlimited. What's the concept behind that?

mihails-strasuns-sociomantic committed:
Add higher level scheduler test

mihails-strasuns-sociomantic committed:
Change important task assert to verify

mihails-strasuns-sociomantic committed:
Fix debug trace from TaskPool.d

mihails-strasuns-sociomantic committed:
Change assert to verify in slowtest

It was evaluating expression for its side-effect which is a bad idea with
assert.

mihails-strasuns-sociomantic commented:
I think this is ready for review. Changes from last update:

- new higher level test which I used to manually trace scheduling logic
- bunch of small fixes everywhere

Testing with some real application would be nice but I don't think it is important enough right now to distract app maintainers (as the effort is non-trivial).

gavin-norman-sociomantic commented on src/ocean/task/Scheduler.d:
```diff
@@ -27,7 +27,6 @@ import ocean.transition;
 import ocean.core.Enforce;
 import ocean.core.Verify;
 import ocean.core.Traits;
-import ocean.core.array.Mutation : reverse;
```
Commit message comment:
"This is the inherent part of running task, not matter the context it is being run from."
->
"This is an inherent part of running a task, not a matter for the context it is being run from."
(Assuming I understand what you want to say.)

gavin-norman-sociomantic commented on src/ocean/task/Task.d:
```diff
@@ -456,8 +456,39 @@ public abstract class Task : ISuspendable
+        {
             this.state_bitmask |= TaskState.Finished;
 
+            if (this.termination_hooks.length)
```
This seems like a potentially dangerous change: running the termination hooks inside the task, rather than from the calling context.

mihails-strasuns-sociomantic replied:
There is no change, it already works like that (the fiber is the same).

gavin-norman-sociomantic replied:
Ah I see, `runTask` is called from inside the fiber. :+1:

gavin-norman-sociomantic commented on src/ocean/task/internal/FiberPoolWithQueue.d:
```diff
@@ -0,0 +1,241 @@
+        Constructor
+
+        Params:
+            stack_size = fiber stack size to use in this poll
```
Incomplete params doc.

gavin-norman-sociomantic replied:
Still incomplete.

mihails-strasuns-sociomantic replied:
https://github.com/sociomantic-tsunami/ocean/pull/293/files#diff-2c0e70ae37423c24132186053ded1353R68

gavin-norman-sociomantic replied:
Ah, the comment fix was in the wrong commit.

gavin-norman-sociomantic commented on src/ocean/task/internal/FiberPoolWithQueue.d:
```diff
@@ -0,0 +1,241 @@
+        debug_trace("running task <{}> via worker fiber <{}>",
+            cast(void*) task, cast(void*) fiber);
+        task.assignTo(fiber);
+        // trick task by using different method for real entry point:
```
This comment is a little mysterious...

gavin-norman-sociomantic commented on src/ocean/task/Scheduler.d:
```diff
@@ -69,6 +70,16 @@ struct SchedulerConfiguration
+    /// it to more than a few dedicated extra pools
+    PoolDescription[] specialized_pools;
+
+    /// specialized worker fiber pool description. Tasks handled by such / pools
```
The "/" at the end of this line seems erroneous.

gavin-norman-sociomantic commented on src/ocean/task/Task.d:
```diff
@@ -83,6 +83,17 @@ public class WorkerFiber : core.thread.Fiber
 
     /***************************************************************************
 
+        Terminates itself by killing underlying task
```
This seems like it should be a separate commit.

gavin-norman-sociomantic commented on src/ocean/task/internal/SpecializedPools.d:
```diff
@@ -0,0 +1,124 @@
+/*******************************************************************************
+
```
Module description missing.

gavin-norman-sociomantic commented on src/ocean/task/internal/SpecializedPools.d:
```diff
@@ -0,0 +1,124 @@
+public class SpecializedPools
+{
+    /// mapping as a simple array (expected to contain few elements)
+    private SpecializedPool[] mapping;
```
What's the reason you're not using an AA?

mihails-strasuns-sociomantic replied:
Cons: slower, inherent heap fragmentation, have to keep in mind all const-related AA bugs
Pros: nice syntax

I think for purely private entity in stdlib choice is obvious :P

gavin-norman-sociomantic commented on src/ocean/task/internal/SpecializedPools.d:
```diff
@@ -0,0 +1,124 @@
+    /***************************************************************************
+
+        Constructor
+
```
Params

gavin-norman-sociomantic commented on src/ocean/task/internal/SpecializedPools.d:
```diff
@@ -0,0 +1,124 @@
+                "Configured stack size is suspiciously low"
+            );
+
+            this.mapping ~= SpecializedPool(
```
Using an AA would also allow us to detect multiple mappings for the same task type.

gavin-norman-sociomantic commented on src/ocean/task/internal/SpecializedPools.d:
```diff
@@ -0,0 +1,124 @@
+
+        Runs task in one of dedicated worker fiber pools if it is registered in
+        the mapping. Returns 'false' otherwise.
+
```
Params / returns.

gavin-norman-sociomantic commented on src/ocean/task/internal/SpecializedPools.d:
```diff
@@ -0,0 +1,124 @@
+
+    ***************************************************************************/
+
+    void kill ( )
```
Protection attribute missing on all methods. (All public, I suppose.)

gavin-norman-sociomantic commented on src/ocean/task/internal/SpecializedPools.d:
```diff
@@ -0,0 +1,124 @@
+    }
+}
+
+private void debug_trace ( T... ) ( cstring format, T args )
```
Unused?

gavin-norman-sociomantic reviewed (commented):
Looking good!

gavin-norman-sociomantic commented:
Oh, needs release notes, too.

mihails-strasuns-sociomantic committed:
Make posible to kill a task via worker fiber

mihails-strasuns-sociomantic commented:
Updated: https://gist.github.com/mihails-strasuns-sociomantic/ce7257baa7584449dce0c21be29d0a7e

mihails-strasuns-sociomantic commented:
I am also bumping priority because investigation on https://github.com/sociomantic-tsunami/ocean/pull/335 shows there are pretty big changes that need to be done on scheduler - and I'd prefer to do those on top of refactoring from this PR.

mihails-strasuns-sociomantic commented:
Added release notes

mihails-strasuns-sociomantic commented:
Green

gavin-norman-sociomantic commented on relnotes/fiber-pool.feature.md:
```diff
@@ -0,0 +1,30 @@
+
+  Such functionality is intended to be used when there are some task types that
+  need real-time handling and thus can't afford waiting in the queue. It can
+  also be used for case when some tasks are very long-living would hold worker
```
Grammar / clarity: "It can also be used for the case where some very long-lived tasks would occupy worker fibers from the shared pool forever, if processed normally."

gavin-norman-sociomantic reviewed (commented):
A couple of minor comments.

mihails-strasuns-sociomantic commented:
Updated:
- relnotes wording
- move doc fix to proper commit

gavin-norman-sociomantic commented:
Feel free to auto-merge when green.

daniel-zullo commented on src/ocean/task/internal/FiberPoolWithQueue.d:
```diff
@@ -0,0 +1,249 @@
+
+    **************************************************************************/
+
+    this ( size_t queue_limit, size_t stack_size, size_t limit,
```
missing protection attribute

daniel-zullo commented on src/ocean/task/internal/FiberPoolEager.d:
```diff
@@ -0,0 +1,83 @@
+
+    **************************************************************************/
+
+    this ( size_t stack_size )
```
missing protection attribute

daniel-zullo commented on src/ocean/task/internal/FiberPoolEager.d:
```diff
@@ -0,0 +1,83 @@
+    import ocean.io.Stdout;
+
+/// Ditto
+class FiberPoolEager : FiberPool
```
missing protection attribute

daniel-zullo commented on src/ocean/task/Scheduler.d:
```diff
@@ -299,12 +318,15 @@ final class Scheduler
-            this.queued_tasks.length(), this.suspended_tasks.length());
+        debug_trace(
+            "Shutting down initiated. {} queued tasks will be " ~
+                " discardead, {} suspended tasks will be killed",
```
typo *discardead

daniel-zullo commented on src/ocean/task/Task.d:
```diff
@@ -299,8 +310,8 @@ public abstract class Task : ISuspendable
     public void suspend ( )
     {
         assert (this.fiber !is null);
```
shouldn't it be replaced to `verify()` too?

mihails-strasuns-sociomantic replied:
No, `task` package is currently the only one which uses both as intended (depending on what kind of condition is protected against), contrary to rest of ocean which was converted 1-to-1.

I will likely have a full re-read of all scheduler source code soon and will also re-evaluate all assert/verify usage cases.

daniel-zullo replied:
ok, good to know that

daniel-zullo commented on src/ocean/task/internal/FiberPoolEager.d:
```diff
@@ -0,0 +1,83 @@
+import core.thread : Fiber;
+
+import ocean.meta.types.Qualifiers;
+import ocean.core.Enforce;
```
this import should go on the top of ocean imports

daniel-zullo commented on src/ocean/task/internal/FiberPoolEager.d:
```diff
@@ -0,0 +1,83 @@
+        Constructor
+
+        Params:
+            stack_size = fiber stack size to use in this poll
```
s/poll/pool ?

daniel-zullo commented on src/ocean/task/internal/FiberPoolWithQueue.d:
```diff
@@ -0,0 +1,249 @@
+        Params:
+            queue_limit = max size of task queue, recommened to be at least as
+                large as pool limit
+            stack_size = fiber stack size to use in this poll
```
ditto *poll

daniel-zullo commented on src/ocean/task/internal/SpecializedPools.d:
```diff
@@ -0,0 +1,178 @@
+import ocean.task.Task;
+import ocean.core.Verify;
+import ocean.core.Enforce;
+import ocean.core.Optional;
```
Should the imports be sorted?

mihails-strasuns-sociomantic replied:
No, there is no particular import order requirement for our libs/apps.

daniel-zullo replied:
fair enough

daniel-zullo reviewed (commented):
A bit late for my minor comments but I'd submit them anyway

mihails-strasuns-sociomantic commented:
@daniel-zullo-sociomantic thanks, I will submit separate style cleanup PR
